feat: refactor into self-contained runner

davidcavazos opened the pull request:
This simplifies the setup required for repositories to maintain. The runner is now fully self-contained to run tests consistently while loading and setting up the environment with the `ci-setup.json` file.

The runner supports `lint` and `test` commands.

This now requires the config file to define `lint` and `test` entries. This allows us to consistently run the same commands both locally and in the CI, and we could even have different flags or additional debugging commands defined for different config files (prod vs dev).

Since the script is not in main yet, to run it via the alias it must point to this PR branch.

```sh
alias custard="curl -sSL https://raw.githubusercontent.com/GoogleCloudPlatform/cloud-samples-tools/refs/heads/local-runner/scripts/src/custard.ts | node - $@"
```

Each can have `pre`, `run`, and `post` commands. More information in the README.

For example, to configure the nodejs-docs-samples repo:

```jsonc
// .github/config/nodejs.jsonc
{
  // ...
  "commands": {
    "lint": {
      "pre": "npm install",
      "run": "npx gts lint"
    },
    "test": {
      "pre": "npm install",
      "run": [
        "npm install",
        "npm test"
      ]
    }
  },
  // ...
}
```

To run a command:

```sh
custard run <config-file> <command> [args...]
```

For example, to run the lint and tests on the nodejs-docs-samples `tpu` package.

```sh
custard .github/config/nodejs.jsonc lint tpu
custard .github/config/nodejs.jsonc test tpu
```

You can pass one or more packages to both `lint` and `test`.

Prerequisites:
- `node` >= 23 for typescript support

briandorsey commented:
Design question: the new config keys (`test`, `lint`, etc)... would the CI actions also be updated to use the same configs? (perhaps by just running this script?)

davidcavazos commented:
> Design question: the new config keys (`test`, `lint`, etc)... would the CI actions also be updated to use the same configs? (perhaps by just running this script?)

Yes, that's the idea. The Makefile would be updated to use this script, pointing to the config file where the test and lint commands are defined. So the both running the Makefile on the CI actions or locally, would both run the same config commands.

briandorsey commented on run.sh:
```diff
@@ -0,0 +1,111 @@
+# limitations under the License.
+
+# Usage:
+#   sh run.sh lint path/to/config.jsonc path/to/package
```
This is recommending `sh`, but towards the end of the file, this script is executing `bash`. Would it make sense to recommend people run this script in `bash` for consistency? I don't know shell scripting well, so this is not a leading question, just wondering.

davidcavazos replied:
It works on either sh or bash. However, sh is available on almost (if not all) *nix environments, while bash is not available on some minimal OS images. Not a big preference, this could be changed since it's probably safe to assume bash is available.

briandorsey commented on run.sh:
```diff
@@ -0,0 +1,111 @@
+#   sh -x run.sh lint path/to/config.jsonc path/to/package
+#   sh -x run.sh test path/to/config.jsonc path/to/package
+
+set -e # Exit immediately if a command exits with a non-zero status.
```
Thanks for this comment, I wondered about this while reading the script.

briandorsey commented:
Design question: Regarding "The test command loads the ci-setup.json file and merges it on top of the defaults defined by the config file. It exports the environment variables, that can be overridden if that environment variable is already set in the local environment. It fetches and exports secrets, also overridden if they're defined locally."

I'm not sure I'm reading this correctly... but I *think* it may be doing something different than I expected with ENV variables. I imagined that there would be a kind of hierarchy of setting environment variables:

```
locally set ENV in shell
set in package `ci-setup.json`
set in repo config
```
... with things higher in that list overriding lower levels. Is that how it works? (I don't think so from reading the comment and the code.) 

If this script overwrites user set shell ENV variables, how will users set things up for their environment? Different auth, or resource names, or project, etc. 

In the script, I see that PROJECT_ID is already expected to exist, though? How does that interact with a project_id set in the configs?

glasnt commented on run.sh:
```diff
@@ -0,0 +1,111 @@
+
+set -e # Exit immediately if a command exits with a non-zero status.
+
+if [ -z "$PROJECT_ID"]; then
```
```suggestion
if [ -z "$PROJECT_ID" ]; then
```

glasnt commented on run.sh:
```diff
@@ -0,0 +1,111 @@
+
+# Print versions to make sure dependencies are installed.
+jq --version
+gcloud --version
```
For some reason on my system this command take 3.5s. Unsure if it's the same on yours; would a `which gcloud` be a valid alternative here?

davidcavazos replied:
Good point, yes that would work

glasnt commented on run.sh:
```diff
@@ -0,0 +1,111 @@
+      fi
+    done
+
+    set -x # Print commands as they are executed.
+    bash -x -c "$(jq -r '."test-pre" // ""' <<< $config)"
+    cd "$package_path"
+    bash -x -c "$(jq -r '.test // ""' <<< $config)"
```
`test-pre` and `test` are not currently defined in the example config in /custard/README.md, nor in the nodejs.jsonc in nodejs-docs-samples. Currently the test command is hardcoded in the custard-ci.yml workflow. 

(Same with lint-pre and lint)

davidcavazos replied:
Yeah, I was testing with my local Node repo, so I modified that config file. I haven't made or modified any tests for this yet.

davidcavazos commented:
> ```
> locally set ENV in shell
> set in package `ci-setup.json`
> set in repo config
> ```

Yes, that's the priority. The script creates a merged version where the ci-setup.json file overrides whatever is defined in the config. So it takes the config defaults as a base, if anything is defined in the ci-setup it takes that, otherwise it takes the defaults. Then, before exporting the variable, if it's already defined locally it uses that, otherwise it uses whatever is defined in the merged config. So the priority is "local env > ci-setup > repo config".

> If this script overwrites user set shell ENV variables, how will users set things up for their environment? Different auth, or resource names, or project, etc.

In retrospective, it would have been clearer to use an if statement. It's using the `${locally_defined:-default_value}` syntax, and then it always exports it. But if it's locally defined, it basically re-exports with the same value. An if statement would have been clearer.

> In the script, I see that PROJECT_ID is already expected to exist, though? How does that interact with a project_id set in the configs?

True, that will not be overridden by the script. That should be fixed to keep how the current one works. Now that I think about it, it's also missing exporting the RUN_ID and variable interpolation.

briandorsey commented:
Ahh, thank you, that makes sense. I completely read through that assignment syntax without realizing it was preserving local values (can you tell I don't know shell scripting well?)

glasnt commented on run.ts:
```diff
@@ -0,0 +1,258 @@
+  }
+
+  default: {
+    console.error(`Unknown command: ${command}`);
```
If you execute `node run.ts` without args you get

```
Unknown command: undefined
usage: node run.ts [lint | test] [options]
```
The response shouldn't return "undefined", just the usage (to match the feel of other script executions, like bash)

glasnt commented on run.ts:
```diff
@@ -0,0 +1,258 @@
+  console.log(`not compatible with ${major}.${minor}`);
+  return false;
+}
+if (checkVersion("make --version", 3, 82)) {
+  make = "make";
+  console.log(`Using ${execSync("which make").toString().trim()}`);
+} else if (checkVersion("gmake --version", 3, 82)) {
+  make = "gmake";
+  console.log(`Using ${execSync("which gmake").toString().trim()}`);
+}
```
What's the reason for requiring make/gmake of this version? 

My first execution of this script came with a lot of this debugging output and I was initially confused.

glasnt commented on run.ts:
```diff
@@ -0,0 +1,258 @@
+function lint(packagePath: string) {
+  const cmd = `${make} lint dir=${packagePath}`;
+  console.log(`>> ${cmd}`);
+  execSync(cmd, { stdio: "inherit" });
```
Tried `node run.ts lint (path-to-code)`, and this failed with `gmake: *** No rule to make target 'lint'.  Stop.`. 

Tried this for nodejs-docs-samples/compute, and just nodejs-docs-samples. 

Unsure if the Makefile being run is the makefile in the cloud-samples-tools repo, or the one in the path.

glasnt replied:
Even after fa0ddbf76815ef2c8efb4e37d1239f572a5f8800 this fails (that commit was for the test not lint). Comment stads

davidcavazos replied:
Sorry for the lack of documentation. The script is meant to be run on the root directory of the langauge repo (not the tools repo) so for example:

`nodejs-docs-samples$ node run.ts test compute`

We still need to figure out and document the recommended way to get and run the script. We could just `curl <script-url> | node` to run it locally, and the GitHub action could checkout the specific file from the specific tag we want.

glasnt reviewed (commented):
Tried the new run.ts version with mixed results

davidcavazos commented:
I'm restructuring this into its own package, adding tests, and testing workflows. The usage should still remain downloading a single file and running directly, but the package allows us to lint and test the code better. After this, I'll adapt the workflow tests to run the local runner instead of the `setup-custard` action.

I'll change this to a draft PR until changes are ready and tests are fixed.

briandorsey commented on scripts/src/custard.ts:
```diff
@@ -0,0 +1,530 @@
+    // Only throw an error if running the script directly.
+    // Otherwise, this file is being imported (for example, on tests).
+    if (process.argv[1] && process.argv[1].match(/custard\.(ts|js)$|^-$/)) {
+      throw new Error(usage('[lint | test] [options]'));
```
I get this warning when running under node v23:

```
> node --trace-warnings src/custard.ts
usage: node custard.ts [lint | test] [options]
(node:78683) ExperimentalWarning: Type Stripping is an experimental feature and might change at any time
    at emitExperimentalWarning (node:internal/util:285:11)
    at stripTypeScriptModuleTypes (node:internal/modules/typescript:166:5)
    at ModuleLoader.<anonymous> (node:internal/modules/esm/translators:551:16)
    at #translate (node:internal/modules/esm/loader:534:12)
    at ModuleLoader.loadAndTranslate (node:internal/modules/esm/loader:581:27)
    at async ModuleJob._link (node:internal/modules/esm/module_job:116:19)
```

davidcavazos replied:
Yes, that warning still shows up in Node 23. From looking at the history, Node 22 introduced the feature under an experimental flag, then 23 removed the experimental flag and made it the default, but looks like the warning is still there. It's not a big issue and it still works as expected.

If we don't want that warning to show up, we can have a built `.js` file and a check that makes sure the js file is up to date. Later on we could have the js file as part of the release published artifacts, but that seems to be a bit trickier to fetch from workflows, so having the js file in git might be the simplest option.

davidcavazos replied:
This warning can be disabled with `export NODE_OPTIONS="--no-warnings=75058"`

glasnt commented:
I tried using the instructions from the README with the alias from the PR description, I can't get a simple lint working running the suggested command on the main branch on nodejs-docs-samples. I'm guessing there's a lack of ci-setup.json file in that repo. 

I'm also guessing the alias config may affect the argv[n] calls in the script, as `custard` vs `node custard.ts` may affect the index? (unless `| node -` affects the arg list)

davidcavazos commented:
You're right, I changed the way the command works and updated the usage, but not the README. I'll update that too.

I'll also  include the changes needed on the config file for the Node repo, that way you can test it.

On how the script is called, like `node custard.ts [args]` vs `.. | node - [args]` does affect how parameters are passed. I made sure both ways work, so it should work the same way. It's also different when it's imported (like for tests), so I also made sure that worked as well.

davidcavazos commented:
Just updated the PR description and the README.

glasnt reviewed (approved):
Approval based on getting the changes in, but as discussed, after this is released we should seek folks to try out the new interface and address any raised usability issues separately.
